### PR TITLE
fix: adjust permissions on safe-settings working dir

### LINKED ISSUE #362: Permission denied when building Docker image
### What is actually happening

When running docker-compose I get an error that the build cannot access `/opt/safe-settings` when installing the npm packages.

I think the issue has been introduced with the recent root/user changes in the Dockerfile. In my opinion the fix is to run the build process as root and change only after the `npm install` command to the node user.

### What is the expected behavior

Building the Docker image should be successful.

### Error output, if available

```
npm WARN deprecated @types/pino-pretty@5.0.0: This is a stub types definition. pino-pretty provides its own type definitions, so you do not need this installed.
npm WARN deprecated @types/pino-std-serializers@4.0.0: This is a stub types definition. pino-std-serializers provides its own type definitions, so you do not need this installed.
npm WARN checkPermissions Missing write access to /opt/safe-settings
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@^2.3.2 (node_modules/jest-haste-map/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.3.2: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})
npm WARN safe-settings@0.1.0-rc.26 No description

npm ERR! code EACCES
npm ERR! syscall access
npm ERR! path /opt/safe-settings
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, access '/opt/safe-settings'
npm ERR!  [Error: EACCES: permission denied, access '/opt/safe-settings'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'access',
npm ERR!   path: '/opt/safe-settings'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_logs/2022-12-13T15_59_49_444Z-debug.log
```

### Are you using the hosted instance of probot/settings or running your own?
self-hosted

### If running your own instance, are you using it with github.com or GitHub Enterprise?
GitHub.com

#### Version of probot/settings
latest

### PR DESCRIPTION
**Pull request by martinm82**
Fixes #362

### CODE REVIEW

**Comment by martinm82** [comment]
@decyjphr could you please have a look at this PR. Thanks

File: Dockerfile [hunk 1]
```diff
@@ -5,6 +5,9 @@ LABEL version="1.0" \
       description="Probot app which is a modified version of Settings Probot GitHub App" \
       maintainer="GitHub Professional Services <services@github.com>"
 
+RUN   mkdir /opt/safe-settings \
+  &&  chown -R node:node /opt/safe-settings
+
 USER node
 ## Set our working directory
 WORKDIR /opt/safe-settings
```
Comment (decyjphr) [inline Dockerfile]: Yes, this should be enough to create the /op/safe-settings dir. We don't need to explicitly create it like in the previous version.

**Review by decyjphr — APPROVED** [review]
👍🏾

File: Dockerfile [hunk 2]
```diff
@@ -5,6 +5,9 @@ LABEL version="1.0" \
       description="Probot app which is a modified version of Settings Probot GitHub App" \
       maintainer="GitHub Professional Services <services@github.com>"
 
+RUN   mkdir /opt/safe-settings \
```
Comment (decyjphr) [inline Dockerfile]: @martinm82 could we make the Dockerfile simpler as follows:
```dockerfile
FROM node:14.17.6-alpine
WORKDIR /opt/safe-settings
ENV NODE_ENV production
## Set the Labels
LABEL version="1.0" \
      description="Probot app which is a modified version of Settings Probot GitHub App" \
      maintainer="GitHub Professional Services <services@github.com>"

## These files are copied separately to allow updates
## to the image to be as small as possible
COPY  package.json /opt/safe-settings/
COPY  index.js /opt/safe-settings/
COPY  lib /opt/safe-settings/lib

## Install the app and dependencies
RUN npm install

## This app will listen on port 3000
EXPOSE 3000

USER node

## This does not start properly when using the ['npm','start'] format
## so stick with just calling it outright
CMD npm start
```

Comment (decyjphr) [reply Dockerfile]: @martinm82 let me know if this above change would work for you. I am not sure why the previous docker file had python install, etc. I don't believe it needs to be that complex.

Probably, we can take this opportunity and update the node version to 18

**Review by quike — APPROVED** [review]

**Comment by freef4ll** [comment]
Would be good to have these merged, as also running into permission issue after an upgrade.

**Comment by decyjphr** [comment]
@martinm82 can we make these changes? Apologies if you have already made those...
```dockerfile
FROM node:14.17.6-alpine
WORKDIR /opt/safe-settings
ENV NODE_ENV production
## Set the Labels
LABEL version="1.0" \
      description="Probot app which is a modified version of Settings Probot GitHub App" \
      maintainer="GitHub Professional Services <services@github.com>"

## These files are copied separately to allow updates
## to the image to be as small as possible
COPY  package.json /opt/safe-settings/
COPY  index.js /opt/safe-settings/
COPY  lib /opt/safe-settings/lib

## Install the app and dependencies
RUN npm install

## This app will listen on port 3000
EXPOSE 3000

USER node

## This does not start properly when using the ['npm','start'] format
## so stick with just calling it outright
CMD npm start
```

**Comment by martinm82** [comment]
> @martinm82 can we make these changes? Apologies if you have already made those...
> 
> ```dockerfile
> FROM node:14.17.6-alpine
> WORKDIR /opt/safe-settings
> ENV NODE_ENV production
> ## Set the Labels
> LABEL version="1.0" \
>       description="Probot app which is a modified version of Settings Probot GitHub App" \
>       maintainer="GitHub Professional Services <services@github.com>"
> 
> ## These files are copied separately to allow updates
> ## to the image to be as small as possible
> COPY  package.json /opt/safe-settings/
> COPY  index.js /opt/safe-settings/
> COPY  lib /opt/safe-settings/lib
> 
> ## Install the app and dependencies
> RUN npm install
> 
> ## This app will listen on port 3000
> EXPOSE 3000
> 
> USER node
> 
> ## This does not start properly when using the ['npm','start'] format
> ## so stick with just calling it outright
> CMD npm start
> ```

Sorry that it took so long. Yes, moving workdir to top seems to solve the issue. Thanks @decyjphr for the better solution.

File: Dockerfile [hunk 3]
```diff
@@ -30,9 +27,6 @@ COPY  lib /opt/safe-settings/lib
 USER root
 RUN apk add --no-cache make python
```
Comment (decyjphr) [inline Dockerfile:28]: We don't need to run install python any more. So these 2 lines can be removed. It'll make the Dockerfile simpler

**Review by decyjphr — CHANGES_REQUESTED** [review]

**Comment by robandpdx** [comment]
@martinm82 @decyjphr I have create #453 to supercede this PR. I've test the container and it seems to be working fine.

**Comment by martinm82** [comment]
Closing in favour of https://github.com/github/safe-settings/pull/453

**Pull request closed without merging**